dll/win32/setupapi/query.c

```diff
@@ -693,3 +693,373 @@ BOOL WINAPI SetupQueryInfOriginalFileInformationW(
+    }
+    else{
+        HKEY handle;
+        LSTATUS status = 0;
+        UINT len = 0;
+        BYTE buffer[MAX_PATH]; // size?
+        BYTE* currentPos = NULL;
+        currentPos = buffer;
+        for(int i=0; i < count; i++){
+            strcpy(currentPos,list[i]);
+            currentPos += strlen(list[i]) + 1;
+        }
+        len = currentPos - buffer;
+
+        if(flags & SRCLIST_SYSTEM){
+            status = RegOpenKeyExA(HKEY_LOCAL_MACHINE,"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Setup",0,KEY_SET_VALUE,&handle);
+            if(status)
+                return FALSE;
+            status = RegSetValueExA(handle,"Installation Sources",0,REG_MULTI_SZ,buffer,len);
+            if(status)
+                return FALSE;
+            RegCloseKey(handle);
+        }
+
+        if(flags & SRCLIST_USER){
+            status = RegOpenKeyExA(HKEY_CURRENT_USER,"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Setup",0,KEY_SET_VALUE,&handle);
+            if(status)
+                return FALSE;
+            status = RegSetValueExA(handle,"Installation Sources",0,REG_MULTI_SZ,buffer,len);
+            RegCloseKey(handle);
+            if(status)
+                return FALSE;
+        }
+    }
+
+    if(flags & SRCLIST_NOBROWSE)
+        noBrowse = TRUE;
+
+    return TRUE;
+}
+
+/***********************************************************************
+ *      SetupSetSourceListW (SETUPAPI.@)
+ */
+BOOL WINAPI SetupSetSourceListW(DWORD flags, PCWSTR *list, UINT count)
+{
+    TRACE("(%X, %p, %d)\n", flags, list, count);
+
+    if(flags & SRCLIST_TEMPORARY){
+        srclist_temporary_sources = (PVOID*)*list;
+        srclist_temporary_sources_count = count;
+    }
+    else{
+        HKEY handle = NULL;
+        LSTATUS status = 0;
+        UINT len = 0;
+        BYTE buffer[MAX_PATH];
+        BYTE* currentPos = NULL;
+
+        currentPos = buffer;
+        for(int i=0; i < count; i++){
+            wcscpy((wchar_t*)currentPos,list[i]);
+            currentPos += wcslen(list[i]) + 1;
+        }
+        len = currentPos - buffer;
+
+        if(flags & SRCLIST_SYSTEM){
+            status = RegOpenKeyExW(HKEY_LOCAL_MACHINE,L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Setup",0,KEY_SET_VALUE,&handle);
+            if(status)
+                return FALSE;
+            status = RegSetValueExW(handle,L"Installation Sources",0,REG_MULTI_SZ,buffer,len);
+            if(status)
+                return FALSE;
+            RegCloseKey(handle);
+        }
+
+        if(flags & SRCLIST_USER){
+            status = RegOpenKeyExW(HKEY_CURRENT_USER,L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Setup",0,KEY_SET_VALUE,&handle);
+            if(status)
+                return FALSE;
+            status = RegSetValueExW(handle,L"Installation Sources",0,REG_MULTI_SZ,buffer,len);
+            RegCloseKey(handle);
+            if(status)
+                return FALSE;
+        }
+    }
+
+    if(flags & SRCLIST_NOBROWSE)
+        noBrowse = TRUE;
```

Comment:
Is this the end of file? You forgot to add a closing brace `}` at the end of the function, so it will not compile properly.

Reply:
Ok, thanks - the patch ate the last two lines.